modules/new_serverpod_auth/serverpod_auth_email_account/serverpod_auth_email_account_server/lib/src/business/email_accounts.dart

```diff
@@ -0,0 +1,458 @@
+      resetToken: resetToken,
+    );
+
+    return PasswordResetResult.passwordResetSent;
+  }
+
+  /// Returns the auth user ID for the successfully changed password
+  static Future<UuidValue> completePasswordReset(
+    final Session session, {
+    required final String resetCode,
+    required final String newPassword,
+  }) async {
+    final resetRequest = (await EmailAccountPasswordResetRequest.db.find(
+            session,
+            where: (final t) => t.verificationCode.equals(resetCode)))
+        .where((final r) => r.created
+            .add(EmailAccountConfig.current.passwordResetCodeLifetime)
+            .isAfter(DateTime.now()))
+        .singleOrNull;
+
+    if (resetRequest == null) {
+      throw Exception('Password reset request not found.');
+    }
+
+    await EmailAccountPasswordResetRequest.db.deleteRow(session, resetRequest);
+
+    final account = (await EmailAccount.db.findById(
+      session,
+      resetRequest.authenticationId,
+    ))!;
+
+    await EmailAccount.db.updateRow(
+      session,
+      account.copyWith(
+        passwordHash: PasswordHash.createHash(
+          email: account.email,
+          password: newPassword,
+        ),
+      ),
+    );
+
+    return account.authUserId;
+  }
+
+  static Future<bool> _hasTooManyFailedSignIns(
+    final Session session,
+    final String email,
+  ) async {
+    final failedLoginAttemptCount =
+        await EmailAccountFailedLoginAttempt.db.count(
+      session,
+      where: (final t) =>
+          t.email.equals(email) &
+          (t.attemptedAt >
+              DateTime.now().toUtc().subtract(
+                  EmailAccountConfig.current.emailSignInFailureResetTime)),
+    );
+
+    return failedLoginAttemptCount >=
+        EmailAccountConfig.current.maxAllowedEmailSignInAttempts;
+  }
+
+  static Future<void> _logFailedSignIn(
+    final Session session,
+    final String email,
+  ) async {
+    await EmailAccountFailedLoginAttempt.db.insertRow(
+      session,
+      EmailAccountFailedLoginAttempt(
+        email: email,
+        ipAddress: session.remoteIpAddress,
+      ),
+    );
+  }
+
+  static Future<EmailAccount?> _importExistingUser(
+    final Session session, {
+    required final String email,
+    required final String password,
+  }) async {
+    final importFunc = EmailAccountConfig.current.existingUserImportFunction;
+
+    if (importFunc == null) {
+      return null;
+    }
+
+    final userId = await importFunc(
+      session,
+      email: email,
+      password: password,
+    );
+
+    if (userId == null) {
+      return null;
+    }
+
+    return await EmailAccount.db.insertRow(
+      session,
+      EmailAccount(
+        authUserId: userId,
+        email: email,
+        passwordHash: password,
+      ),
+    );
+  }
+
+  /// Cleans up the log of failed login attempts older than [olderThan].
+  static Future<void> deleteFailedLoginAttempts(
+    final Session session, {
+    required final Duration olderThan,
```

Comment:
TODO: When the "check time" is configurable, we could make this optional and automatically delete items older than what would be checked on subsequent requests.

Reply:
This has been implemented and `olderThan` is now optional.